specs/Standards/Tokens/NonFungibleToken.md

```diff
@@ -0,0 +1,167 @@
+- Initialize contract once. The given total supply will be owned by the given account ID. (Not in template).
+
+There are a few concepts in the scenarios above:
+
+- **Total supply**. The total number of tokens in circulation.
+- **Token owner**. An account ID that owns one or more tokens.
+- **Transfer**. Action that moves some amount from one account to another account.
+- **Escrow**. A different account from the balance owner who has permission to one or more tokens.
+- **Access**. The specific token ID that another account has access to.
+
+### **Simple transfer**
+
+#### **Assumptions**
+
+- the Corgi nft contract is `corgi`
+- Alice's account is `alice`
+- Jeraldo's account is `jerry`
+- The NFT contract has been initialized with a non-zero token supply
+- There exists a token with the ID of `3`
+
+#### **High-level**
+
+Alice needs to issue one transaction to the Corgi NFT contract to transfer one corgi token to Jeraldo.
+
+#### **Technical calls**
+
+1. `alice` calls `corgi::transfer({"new_owner_id":"jerry", "token_id":3})`
+
+### **Token swap through a third party escrow**
+
+Alice wants to transfer one Corgi NFT through a third party escrow to Jeraldo in exchange for one Sausage NFT. NOTE: This standard does not include how escrow handles anything outside of the explicit transfer of tokens. Any call on an escrow account is illustrative, but not standardized. Escrow can be a trusted individual who acts as an intermediary.
+
+#### **Assumptions**
+
+- the Corgi nft contract is `corgi`
+- the Sausage nft contract is `sausage`
+- Alice's account is `alice`
+- Jeraldo's account is `jerry`
+- The Escrow contract is `escrow`
+- The NFT contract has been initialized with a non-zero token supply
+- There exists a Corgi token with the ID of `3` and a Sausage token with the ID of `5`
+- The Escrow contract manages how the transfer is facilitated and guarantees requirements are met for transfer
+
+#### **High-level**
+
+Both Alice and Jerry will issue asynchronous transactions to their respective contracts, `corgi` and `sausage` to grant access to the escrow to trade tokens on their behalf. `escrow` will call the `sasuage` token contract asynchrounously to transfer the Sausage token to `escrow`. After, `escrow` will also call the `corgi` contract to asynchornously transfer the Corgi token to `escrow`. Then, `escrow` will conduct a transfer to both parties.
+
+- If both of the `transfer` calls succeed, then Alice will now own one Sausage token and Jerry will own one Corgi token.
+- If one or both of the final `transfer` calls fail, then nothing will happen and `escrow` should attempt reissuing the failed transaction.
+
+#### **Technical calls**
+
+1. `alice` makes an async call to `corgi::grant_access({"escrow_account_id":"escrow"})`
+2. `jerry`  makes an async call to `sausage::grant_access({"escrow_account_id":"escrow"})`
+3. `escrow` calls `sausage::transfer({"new_owner_id:"escrow", "token_id": 5})`
```

Comment:
This might be a security issue. When escrow wants to transfer the token 5 from `alice` by the given `token_id`, the token may not be owned by `alice` anymore. Since `alice` can transfer it to `jerry`. If jerry also gave the access to `escrow` (pretty common for marketplaces), then `escrow` will take the token out of `jerry` and assign ownership to `alice.

The easiest way to address it is to explicitly provide `owner_id` in the `transfer_from` function. So might need to put it back.

@janedegtiareva @potatodepaulo

Reply:
Is escrow malicious in this case, or are we trying to prevent some sort of race condition/working with stale state?

Reply:
> `escrow` will take the token out of `jerry` and assign ownership to `alice`

Assuming a trustworthy escrow, wouldn't this situation imply that the tokens still get where they were intended?

* alice gives access to escrow
* jerry gives access to escrow
* alice & jerry swap directly, without escrow
* escrow performs "swap", but ends up taking target tokens away from jerry & alice just to give them back again

seems like a waste of gas, but like it wouldn't result in a mis-swap _in this particular case_

However, it seems like it might create an attack vector, though I haven't thought of a particular attack scenario yet. Can you think of one?

Reply:
`escrow` is not malicious. So it's somewhat race condition.

`escrow` can't rely on the result of `get_token_owner` to determine the current owner, because by the time it gets the result, the owner may change. This means `alice` can call trade with `jerry` once and then trade the same token with `bob` again. The 2nd time, the token is owned by `bob`, but `escrow` can't know about it.

If `escrow` first calls `get_token_owner` to determine the current owner, then by the time the result got back to `escrow` the owner may already change. So `alice` can trade on `escrow2` with `jerry` in parallel, if `get_token_owner` executed before the `escrow2` trade, then `escrow` thinks that the `alice` still owns the token.

Reply:
There is an easy way to avoid it. Since there is only one `owner` of each token, then we don't need to call `get_token_owner` and `transfer_from` in two different calls. In fact `transfer_from` can call `get_token_owner` inside, and check if the callee is authorized for executing transfer on behalves of the owner. Since all this operations are happening inside the same smart contract (when calling `transfer_from`) then we don't have race condition.

Reply:
We can't remove `owner_id` from transfer_from, because of the async execution. If the 2 owners authorized access to the same `marketplace`, the marketplace will be able to transfer tokens from both owners. So any of these 2 owners can call a method on the marketplace that will call the `transfer_from` assuming the ownership.
The token contract can't verify the who's initiated the transfer because we use immediate predecessor (in this case marketplace) instead of the use signer of the transaction as a proof of ownership.